### PR TITLE
feat(conan): Add conan support

### LINKED ISSUE #948: Add Conan support

### PR DESCRIPTION
**Pull request by TobiaszSML**
This PR introduces Conan package manager to containerbase.
It should close #948 which is part of #21912

closes #1071

### CODE REVIEW

File: src/usr/local/containerbase/tools/v2/python.sh [hunk 1]
```diff
@@ -36,6 +36,10 @@ function prepare_tool() {
   apt_install \
     default-libmysqlclient-dev \
     gcc \
+    clang \
+    make \
+    cmake \
+    perl \
```
Comment (viceice) [inline src/usr/local/containerbase/tools/v2/python.sh:42]: no, if Conan requires them, then add a prepare step to the conan installer

File: src/usr/local/containerbase/tools/v2/conan.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# shellcheck source=/dev/null
+. "$(get_containerbase_path)/utils/python.sh"
+
+function link_tool () {
+  post_install
+  [[ -n $SKIP_VERSION ]] || conan --version
+  conan profile detect
```
Comment (viceice) [inline src/usr/local/containerbase/tools/v2/conan.sh:9]: why this?

**Review by viceice — CHANGES_REQUESTED** [review]
checkout my PR for prepare step 

https://github.com/containerbase/base/pull/1071

File: src/usr/local/containerbase/tools/v2/conan.sh (hunk 2 above)
Comment (TobiaszSML) [reply src/usr/local/containerbase/tools/v2/conan.sh:9]: To properly test Conan, a Conan profile is required. However, the Conan profile depends on the processor architecture and compiler version, and the compiler version varies among Ubuntu releases. So IMHO `conan profile detect` is the easiest and most elegant way to get a Conan profile.

**Review by viceice — COMMENTED** [review]
there are still missing pieces from my PR

https://github.com/containerbase/base/pull/2452#pullrequestreview-1993501823

the profile should be created by prepare step

**Comment by TobiaszSML** [comment]
> there are still missing pieces from my PR
> 
> [#2452 (review)](https://github.com/containerbase/base/pull/2452#pullrequestreview-1993501823)
> 
> the profile should be created by prepare step

Okay, you're in charge here :)
I will add the profile in prepare step, which will depend on Ubuntu distribution and CPU architecture.

I also have one question:
Do I need this NEEDS_PREPARE variable and its associated few lines of code in python.sh?
I'm asking because currently my code has a prepare_tool function and the tests pass without the variable mentioned.

**Comment by viceice** [comment]
> > there are still missing pieces from my PR
> > [#2452 (review)](https://github.com/containerbase/base/pull/2452#pullrequestreview-1993501823)
> > the profile should be created by prepare step
> 
> Okay @viceice, you're in charge here :) I will add the profile in prepare step, which will depend on Ubuntu distribution and CPU architecture.
> 
> I also have one question: Do I need this NEEDS_PREPARE variable and its associated few lines of code in python.sh? I'm asking because currently my code has a prepare_tool function and the tests pass without the variable mentioned.

Yes, it's needed for auto prepare.

File: test/Dockerfile.jammy [hunk 3]
```diff
@@ -125,7 +125,7 @@ RUN install-tool pnpm 9.0.6
 RUN install-tool yarn 1.22.22
 
 #--------------------------------------
-# Image: test-python
+# Image: test-php
```
Comment (viceice) [inline test/Dockerfile.jammy:128]: nice catch, fixed on main.

Please add test to `test/Dockerfile.noble`

File: test/python/Dockerfile [hunk 4]
```diff
@@ -289,6 +307,7 @@ COPY --from=test-poetry-e /.dummy /.dummy
 COPY --from=test-pipenv-a /.dummy /.dummy
 COPY --from=test-pipenv-b /.dummy /.dummy
 COPY --from=test-pdm /.dummy /.dummy
+COPY --from=test-conan /.dummy /.dummy
```
Comment (viceice) [inline test/python/Dockerfile:310]: please sort test alphanumerical

File: test/python/Dockerfile.arm64 [hunk 5]
```diff
@@ -68,3 +76,4 @@ COPY --from=test-checkov /.dummy /.dummy
 COPY --from=test-hashin /.dummy /.dummy
 COPY --from=test-pipenv /.dummy /.dummy
 COPY --from=test-poetry /.dummy /.dummy
+COPY --from=test-conan /.dummy /.dummy
```
Comment (viceice) [inline test/python/Dockerfile.arm64:79]: sort

File: src/usr/local/containerbase/tools/v2/conan.sh [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  esac
+}
+
+function get_distribution() {
+  local distribution
+  distribution="$(get_distro)"
+
+  case $distribution in
+  "focal") echo "$distribution" ;;
+  "jammy") echo "$distribution" ;;
```
Comment (viceice) [inline src/usr/local/containerbase/tools/v2/conan.sh:28]: support `noble`

File: src/usr/local/containerbase/tools/v2/conan.sh [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    "aarch64-focal")
+      create_conan_profile "armv8" "9" > "${USER_HOME}/.conan2/profiles/default"
+      ;;
+    "x86_64-jammy")
+      create_conan_profile "x86_64" "11" > "${USER_HOME}/.conan2/profiles/default"
+      ;;
+    "aarch64-jammy")
+      create_conan_profile "armv8" "11" > "${USER_HOME}/.conan2/profiles/default"
+      ;;
+  esac
```
Comment (viceice) [inline src/usr/local/containerbase/tools/v2/conan.sh:72]: i would prefer two separate switch-case, because this will become more complex with noble

**Review by viceice — COMMENTED** [review]
looks great so far

File: test/Dockerfile.noble [hunk 8]
```diff
@@ -150,6 +152,7 @@ RUN install-tool pdm 2.15.1
 # renovate: datasource=pypi
 RUN install-tool poetry 1.8.2
 
+
```
Comment (viceice) [inline test/Dockerfile.noble:155]:
```suggestion
```

**Comment by TobiaszSML** [comment]
> Please stop merging main if not required, it'll waste time and blocks ci runs 😉

Sure, I will remember that ;)

BTW, I double checked code before merging and noticed that in the Conan profile I was using the standard library with the old ABI. I already pushed commit with the fix for that. Now we can merge this PR. Sorry for this inconvenience.

**Review by viceice — APPROVED** [review]

**Pull request merged**